
syntax = "proto3";

message NullMessage {}

message NIter {
    fixed32 n_iter = 1;
}

message Step {
    fixed64 g = 1;
}

message NdArray {
    string dtype = 1;
    repeated int32 shape = 2;
    bytes data = 3;
}

message FilterState {
    fixed64 n = 1;
    NdArray mean = 2;
    NdArray std = 3;
}

message Experience {
    fixed32 n_iter = 1;
    repeated NdArray observation = 2;
    repeated NdArray action = 3;
    repeated NdArray prob = 4;
    repeated double reward = 5;
    bool terminated = 6;
    fixed32 length = 7;
    FilterState diff = 8;
}

message ScalarMetric {
    string name = 1;
    double y = 2;
    message Arg {
        double value = 1;
    }
    Arg x = 3;
}

service ParameterServer {
    rpc GetGlobalT(NullMessage) returns (Step) {}
    rpc GetFilterState(NullMessage) returns (FilterState) {}
    rpc WaitForIteration(NullMessage) returns (NIter) {}
    rpc SendExperience(Experience) returns (NullMessage) {}
    rpc ReceiveWeights(NIter) returns (stream NdArray) {}
    rpc StoreScalarMetric(ScalarMetric) returns (NullMessage) {}
}
